src/kernels/jupyter/finder/universalRemoteKernelFinder.ts

```diff
@@ -0,0 +1,393 @@
+import { computeServerId } from '../jupyterUtils';
+import { PYTHON_LANGUAGE } from '../../../platform/common/constants';
+import { createPromiseFromCancellation } from '../../../platform/common/cancellation';
+import { DisplayOptions } from '../../displayOptions';
+import * as localize from '../../../platform/common/utils/localize';
+import { isArray } from '../../../platform/common/utils/sysTypes';
+import { deserializeKernelConnection } from '../../helpers';
+import { noop } from '../../../platform/common/utils/misc';
+import { IApplicationEnvironment } from '../../../platform/common/application/types';
+import { KernelFinder } from '../../kernelFinder';
+import { RemoteKernelSpecsCacheKey, removeOldCachedItems } from '../../common/commonFinder';
+import { IContributedKernelFinderInfo } from '../../internalTypes';
+import { disposeAllDisposables } from '../../../platform/common/helpers';
+
+// Even after shutting down a kernel, the server API still returns the old information.
+// Re-query after 2 seconds to ensure we don't get stale information.
+const REMOTE_KERNEL_REFRESH_INTERVAL = 2_000;
+
+// This class watches a single jupyter server URI and returns kernels from it
+export class UniversalRemoteKernelFinder implements IRemoteKernelFinder, IContributedKernelFinderInfo, IDisposable {
+    /**
+     * List of ids of kernels that should be hidden from the kernel picker.
+     */
+    private readonly kernelIdsToHide = new Set<string>();
+    kind: string = 'remote';
+    id: string;
+    displayName: string;
+    private _cacheUpdateCancelTokenSource: CancellationTokenSource | undefined;
+    private cache: RemoteKernelConnectionMetadata[] = [];
+
+    private _onDidChangeKernels = new EventEmitter<void>();
+    onDidChangeKernels: Event<void> = this._onDidChangeKernels.event;
+
+    private _initializeResolve: () => void;
+    private _initializedPromise: Promise<void>;
+
+    private readonly disposables: IDisposable[] = [];
+
+    get initialized(): Promise<void> {
+        return this._initializedPromise;
+    }
+
+    private wasPythonInstalledWhenFetchingKernels = false;
+
+    constructor(
+        private jupyterSessionManagerFactory: IJupyterSessionManagerFactory,
+        private interpreterService: IInterpreterService,
+        private extensionChecker: IPythonExtensionChecker,
+        private readonly notebookProvider: INotebookProvider,
+        private readonly globalState: Memento,
+        private readonly env: IApplicationEnvironment,
+        private readonly cachedRemoteKernelValidator: IJupyterRemoteCachedKernelValidator,
+        kernelFinder: KernelFinder,
+        private readonly kernelProvider: IKernelProvider,
+        private readonly extensions: IExtensions,
+        private isWebExtension: boolean,
+        private readonly serverUri: IJupyterServerUriEntry
+    ) {
+        // Register with remote-serverId as our ID
+        this.id = `${this.kind}-${serverUri.serverId}`;
```

Comment:
Note: class is mostly the same as remoteKernelFinder. Changes to all using serverId of the given serverUri as the key for cache + for registering with the main kernel finder and disposing if removed.